main.nf

```diff
@@ -0,0 +1,519 @@
+
+      script:
+      """
+      samtools faidx $fasta
+      """
+  }
+}
+
+if(!fastagz) {
+  process preprocess_fastagz {
+      publishDir "$baseDir/sampleDerivatives"
+
+      input:
+      file(fasta) from fastaToGzCh
+
+      output:
+      file("*.gz") into (tmpFastaGzCh, fastaGzCh)
+
+      script:
+      """
+      bgzip -c ${fasta} > ${fasta}.gz
+      """
+  }
+}
+
+if(!gzfai) {
+  process preprocess_gzfai {
+    publishDir "$baseDir/sampleDerivatives"
+
+    input:
+    file(fasta) from fastaToGzFaiCh
+    file(fastagz) from tmpFastaGzCh
+
+    output:
+    file("*.gz.fai") into gzFaiCh
+
+    script:
+    """
+    samtools faidx $fastagz
+    """
+  }
+}
+
+if(!gzi){
+  process preprocess_gzi {
+    publishDir "$baseDir/sampleDerivatives"
+
+    input:
+    file(fasta) from fastaToGziCh
+
+    output:
+    file("*.gz.gzi") into gziCh
+
+    script:
+    """
+    bgzip -c -i ${fasta} > ${fasta}.gz
+    """
+  }
+}
+
+//channel to collect all necessary reference files
+fastaChannel = Channel.from(fastaCh).merge(bedCh, faiCh, fastaGzCh, gzFaiCh, gziCh).collect()
+
+/********************************************************************
+  process preprocess_bam
+  Takes care of the read group line.
+********************************************************************/
+
+process preprocess_bam{
+
+  tag "${bam}"
+  publishDir "$baseDir/sampleDerivatives"
+
+  input:
+  file(bam) from bamChannel
+
+  output:
+  set file("ready/${bam}"), file("ready/${bam}.bai") into completeChannel
+
+  script:
+  """
+  mkdir ready
+  [[ `samtools view -H ${bam} | grep '@RG' | wc -l`   > 0 ]] && { mv $bam ready;}|| { picard AddOrReplaceReadGroups \
+  I=${bam} \
+  O=ready/${bam} \
+  RGID=${params.rgid} \
+  RGLB=${params.rglb} \
+  RGPL=${params.rgpl} \
+  RGPU=${params.rgpu} \
+  RGSM=${params.rgsm};}
+  cd ready ;samtools index ${bam};
+  """
+}
+
+/********************************************************************
+  process make_examples
+  Getting bam files and converting them to images ( named examples )
+********************************************************************/
+
+process make_examples{
+
+  tag "${bam}"
+
+  input:
+  set file(fasta), file(bed), file(fai), file(fastagz), file(gzfai), file(gzi) from fastaChannel
+  set file(bam), file(bai) from completeChannel
+
+  output:
+  set val("${bam}"),file('shardedExamples') into examples
+
+  script:
+  """
+  mkdir logs
+  mkdir shardedExamples
+  dv_make_examples.py \
+  --cores ${task.cpus} \
+  --sample ${bam} \
+  --ref ${fastagz} \
+  --reads ${bam} \
+  --regions ${bed} \
+  --logdir logs \
+  --examples shardedExamples
+  """
+}
+/********************************************************************
+  process call_variants
+  Doing the variant calling based on the ML trained model.
+********************************************************************/
+
+process call_variants{
+
+  tag "${bam}"
+
+  input:
+  set val(bam),file('shardedExamples') from examples
```

Comment:
Can use a variable input name here if following above comment. Same as other processes - `file(shardedExamples)` then refer to in script as `$shardedExamples` instead of just `shardedExamples`.